### PR TITLE
Add support for motion v4.4

### PR DESCRIPTION
**Pull request by MichaIng**
Additionally:
- Do partial imports for some modules
- Migrate to f-strings where applicable

ToDo:
- [ ] Inform Home Assistant addon developers once this has been merged: https://github.com/hassio-addons/addon-motioneye/issues/350

### CODE REVIEW

**Comment by MichaIng** [comment]
I need some help guys, the mapping currently contains some dummy conversion functions which need to be finished:
- `motion` v4.3 to/from v4.4 options mapping is a little tricky.
- All `netcam_` options have been merged into a single `netcam_params` options, with comma-separated key-value pairs as value.
- So for `_MOTION_43_TO_44_OPTIONS_MAPPING` we need to loop through `netcam_` options and concatenate the `netcam_params` value with respective key-value pairs. The way the mapping works currently, I'm not sure how to do this, or whether we need to do/change it in the parent function.
- For `_MOTION_44_TO_43_OPTIONS_MAPPING` we need to loop through the `netcam_params` key-value pairs and assign the respective value to the respective `netcam_` option. With splitting this should be possible, but I'd need to check some Python docs and examples 😅.

I hope someone finds time to update/finish this PR accordingly.

**Comment by MichaIng** [comment]
That would be great. Take your time, of course. I'm also busy in new job right now, so don't have the time to look deeper into Python depths for this. Probably on weekend I find some time.

**Comment by zagrim** [comment]
> Shall we instead obtain the motion version a single time at motionEye server start, store it in a global variable and use this for version comparisons? Would reduce overhead and would allow us as well to print a warning if the motion version is below 4.1 or above 4.4, i.e. unsupported, which means that some camera options may not work as expected.

This sounds like a reasonable thing to do. :+1:  It would make the issue of "too new Motion version" more clear and reduce confusion in that particular case. I'm not sure if I've seen an issue ever reported for too old Motion, but sure that also possible to happen.

Indeed the netcam_* stuff seems a bit tricky, and my thin Python experience is not of any help in this.

**Comment by tunkio** [comment]
Really fast implementation of the mapper what I told earlier. It's done with older python which didn't support f-strings or there are no typings but, if you think that would be good way to go I'll do conversions.

https://github.com/tunkio/motioneye/tree/motion_config_mapper

goto utils and run python3 motion_config_mapper.py. That just demonstrates conversions between different versions.

Under the motion folder there are couple example camera configs 4.1, 4.2 and 4.3. Mappings are done from motioneye -> motion version and motion version -> motioneye.

Yaml is used in mapping files so if there is changes later(motion 4.5) it should work so that you just add new mapping file with changed values. 

We could add variable checking there too, now I just listed variable types, defaults and possible values in motioneye.yaml. Didn't implement any verifications yet.

These mappings aren't complete, only for testing.

**Comment by tunkio** [comment]
> Currently the motion version is obtained via external motion -h call for every version comparison.
> This means that now it's called two times whenever camera options are read or set.
> Shall we instead obtain the motion version a single time at motionEye server start, store it in a global variable and use this for version comparisons? Would reduce overhead and would allow us as well to print a warning if the motion version is below 4.1 or above 4.4, i.e. unsupported, which means that some camera options may not work as expected.

There is cache already.(if I'm looking same function 😮)
https://github.com/motioneye-project/motioneye/blob/2b146dd46904c21e90ab87694789ab810416414f/motioneye/motionctl.py#L38-L41

**Comment by tunkio** [comment]
Replaced old mapping implementation with this new one(repo given above) and seems to work somehow. Need to check what breaks and so on. 4.4.0 motion seems to get correct camera conf.

Automatic conversion from older/newer version requires that we add version number attribute into motion and camera configs. Configs might be from older/newer version, so we should use proper mapper to parse those.

**Comment by MichaIng** [comment]
> There is cache already.(if I'm looking same function 😮)

Ah, somehow I've overseen this 😅, then all is fine.

> if you think that would be good way to go I'll do conversions.

Looks actually like a great long-term solution to cover all such cases. I haven't tested or fully understand the YAML handling yet, but looks like you covered the multi-argument values nicely.

I think there are more motion settings covered than actually used in motionEye, isn't it? Shall we reduce this to have mappings only for actually used settings?

Probably we can add the test to be run with `pytest` in CI pipeline: https://github.com/motioneye-project/motioneye/tree/dev/tests

> Didn't implement any verifications yet.

Since we don't have any verification now either, I think it's fine, especially when a CI test compares the generated config with the expected sample.

**Comment by tunkio** [comment]
@MichaIng ok! If it looks ok, I will continue this. I forget the verfication right now and remove those extra parameters. I thought that if we have all parameters there, it would be easy just use them if needed later. But you are right, it might just add more confusion. It's not hard to add the new ones later.

One thing what is not implemented into this test repo, is cache :) I do that too.

I think there might be easy way to do automatic detection of the current camera and motion config version. If we do some version checking for the existing parameters, we can do automatic conversion from old(example 4.2) -> new(4.4) or new(4.4) -> old(4.2) when we start motioneye. This means that you can update motion itself and just restart motioneye and automatically get updated configuration.

I tested this early implementation and in my tests 4.4 worked well, parameters got proper values(list ones), same with 4.2. There might be problems what I didn't see but most things worked as expected.

**Comment by tunkio** [comment]
@MichaIng if you have time 😄 and motivation you could test version below from my repo.

https://github.com/tunkio/motioneye/tree/motion_config_mapper

I added python script for updating existing motion and camera configs. It might take too much time for now to make that automatic. Script is located at motioneye/scripts/update_motion_config_version.py <from_version> <to_version> 

It will update motion.conf and camera-x.conf files in /etc/motioneye/ folder. Automatic backup will be taken into /etc/motioneye/<version>/<datetime>. Downgrade and upgrade possibility.

I didn't remove unused configurations from yaml files yet. We should keep all config differences so we can convert configs between different motion versions.(like 4.1->4.4 or 4.4->4.1). 

Update script uses functions from config.py.

You can run server as well with different motion versions.

**Comment by MichaIng** [comment]
I'd say we need an automated mapping which simply works without manual user interaction, so we can do a motionEye release. I hope I find time to test your implementation, was busy with new job and main OSS project.

One thing I had in mind is that at least the previous/current mapping functions do not cover the case that `motion` has been upgraded while the `camera.conf` still contains settings for the older motion version. Since the `motion -v` result gives the new version then, no config migration is done, so affected settings need to be re-set and saved then. Not sure how this can be handled best, but one idea would be to store the motion version into the `camera.conf` as well, and use this to derive which mapping/migration needs to be done when parsing it.

**Comment by xl442** [comment]
Hi,
I don't know the long term plans here like keeping motion eye option names in sync with (newest) motion or sticking with current motioneye option names and converting back and forth to/from motion names.

From code it looks like final motioneye option names remain at motion 4.3. But just guessing.

Anyway here I supply a patch which works for me at least on v4l and webcam (rtsp) cameras.
Tested on debian bullseye with motion 4.4.0-1

**Comment by MichaIng** [comment]
I implemented a solution for the netcam parameter mapping, hope it works, will test thoroughly before opening for review. Netcam URL, user and password parameters actually do not need to be translated, they are still separate settings in motion 4.4.

**Comment by MichaIng** [comment]
Works great now on all my tests with all possible parameter types. I also updated the `motioneye_init` script to pull motion v4.4 if available. Since neither Debian, not Ubuntu ship motion v4.4 yet, it is now tried to be downloaded whenever the distro code name can be detected (else it is doomed to fail), but gracefully handles the case where no matching package is available or the download fails. E.g. there is no Bullseye package for ARMv6hf (RPi 1 and Zero (1)) available yet. In this case the package from the distro's repository will be installed as fallback.

**Comment by MichaIng** [comment]
@tunkio @zagrim @xl442 
Do you find time for a test/review? I would like to push this, add some more informational output to `motioneye_init` and then push a pre-release (on GitHub + PyPI) until this weekend.

**Comment by zagrim** [comment]
I fear that I won't have time/energy to really test this, but surely I can try to review it (just not right now, I've exhausted my brain for today).

**Review by zagrim — COMMENTED** [review]
Not being that well versed in Python  might have neglected some details, but overall I think it is ok.

While validating the 4.3 to 4.4 mapping I noticed that there's `_USED_MOTION_OPTIONS` which is used for separating out Extra Motion Options. That list has not been updated, so I wonder how well it works anymore with Motion 4.4? In theory some 4.4 options that ME does use might leak to that UI field... :thinking:

File: motioneye/config.py [hunk 1]
```diff
@@ -178,10 +179,86 @@ def webcontrol_interface(v, data):
...
+    elif v == False:
+        v = 'off'
+
+    if 'netcam_params' in data and data['netcam_params']:
+        return {'netcam_params': data['netcam_params'] + ',keepalive = ' + v}
+
+    return {'netcam_params': 'keepalive = ' + v}
+
+
+def netcam_tolerant_check_params(v, data):
```
[the remark this answers is not in the document]
Comment (MichaIng) [reply motioneye/config.py:200]: From 4.3 to 4.4, the individual netcam parameter settings need to be merged into `netcam_params` as multiple posts of its value, hence data needs to be mutated (and only for this reason still passed to the functions) when reading in the individual 4.3 settings.

The other way round the parts of the `netcam_ param` value needs to be split again into the individual settings.

This more complex mapping compared to mapping a single key to another single key was while I asked for help here in the first case 🙂.

**Comment by MichaIng** [comment]
You are right, the extra motion options users may have set (manually) are not translated and were not before. We'd need to have a look how those are applied and do param split and mapping there, but it's probably too much for now?

**Comment by zagrim** [comment]
> You are right, the extra motion options users may have set (manually) are not translated and were not before.

That's one aspect of it, and I agree that it is perhaps too much to try to do complete config migration when we don't really have anything to check the result against.

But my point was that `_USED_MOTION_OPTIONS` list is used by `motion_camera_dict_to_ui` to select what goes to Extra Motion Options, and by `motion_camera_ui_to_dict` to (over)write config with what is in Extra Motion Options. Or did I miss something that prevents, for example, from "video_params" (which is the result of migrating "vid_control_params") from ending up in Extra Motion Options when `motion_camera_dict_to_ui` is next run, since `_USED_MOTION_OPTIONS` contains the old "vid_control_params" but not the new "video_params"? This is of course something that could be checked by testing the migration, which I can't find the time for now.

**Comment by MichaIng** [comment]
Ah, I'll have a closer look at that. Several tests on all directions however went well and I always checked the camera config file before and afterwards.

File: motioneye/config.py [hunk 2]
```diff
@@ -178,10 +179,86 @@ def webcontrol_interface(v, data):
...
+    if 'netcam_params' in data and data['netcam_params']:
+        if v:
+            return {'netcam_params': data['netcam_params'] + ',tolerant_check = on'}
+
+        return {'netcam_params': data['netcam_params'] + ',tolerant_check = off'}
+
+    if v:
+        return {'netcam_params': 'tolerant_check = on'}
+
+    return {'netcam_params': 'tolerant_check = off'}
```
Comment (zagrim) [inline motioneye/config.py:210]:
```suggestion
    if v:
        v = 'on'
    else:
        v = 'off'
    
    if 'netcam_params' in data and data['netcam_params']:
        return {'netcam_params': data['netcam_params'] + ',tolerant_check = ' + v}

    return {'netcam_params': 'tolerant_check = ' + v}
```

File: motioneye/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zagrim) [reply motioneye/config.py:200]: Let me rephrase in an attempt to get my point through, even though it's such a small thing in the end :) 

Why have `netcam_tolerant_check_params` and `netcam_use_tcp_params` been written in different fashion than `netcam_keepalive_params`? All of those three are used for 4.3 -> 4.4 migration, all of those config parameters are to be migrated similarly. Mapping func parameter `v` is used to decide between two values for the v4.4 param, and `data['netcam_params']` (if it exists already) is passed along), yet the function implementation is not the same.
Mostly it is IMO code readability and repetition issue in `netcam_tolerant_check_params` and `netcam_use_tcp_params`. Mutating `v` might not be an real issue at all if it is not used again later by the mapping engine.

To make it clear what I meant, I suggested a change for `netcam_tolerant_check_params` so that it IMO would be more readable (and similar to `netcam_keepalive_params`).

[the remark this answers is not in the document]
Comment (MichaIng) [reply motioneye/config.py:200]: Ah sorry for misunderstanding. `netcam_keepalive_params` was handled differently since it can be `force` as well. But you are right, using the same structure for the others (just `else` instead of `elif`) makes it look cleaner. It can be even a one-liner there.

[the remark this answers is not in the document]
Comment (MichaIng) [reply motioneye/config.py:196]: Done, and all converted to one-liner, so looks much cleaner now.

**Comment by MichaIng** [comment]
I had a closer look into where `_USED_MOTION_OPTIONS` is used. And as far as I can see it is indeed only used after the config dict has been converted to motion v4.3 already, respectively before it gets converted to local motion version. By times some more code comments and probably resorting/factoring would be beneficial, to make it easier for contributors to understand in which function what is done and why it is done in that order/stage of reading/writing configs from/to config files. Also there seem to be quite much redundancy and potential for enhancements, for the future 🙂.

**Review by zagrim — APPROVED** [review]
